numpy/core/src/multiarray/getset.c

```diff
@@ -384,7 +384,8 @@ array_data_set(PyArrayObject *self, PyObject *op)
     }
     if (PyArray_FLAGS(self) & NPY_ARRAY_OWNDATA) {
         PyArray_XDECREF(self);
-        PyDataMem_FREE(PyArray_DATA(self));
```

Comment:
This should have been `npy_free_cache`

Reply:
I think this raises a possible risk in these changes - as I understand it, it didn't matter if the `npy_*_cache` functions were used for the allocation and not the free or vice versa; everything would behave correctly, it would just affect the cache capacity.

With the changes in this PR, either:

* We make the same promise to user hooks, that things you free may be deallocated by someone else or vice versa
* We need to audit every free in our code (and likely, accept bug reports for a couple releases for segfaults due to ones we miss)

I suppose as long as our default allocator is happy to accept raw mallocs and frees, this won't actually result in a regression.